File: src/commands/dep.ts
import { parseArgs } from 'node:util';

import { Ticket } from '../lib/ticket.ts';
import { buildTicketMap, exitWithError } from './utils.ts';

export async function cmdDepTree(args: string[]): Promise<void> {
  const { values, positionals } = parseArgs({
    args,
    options: {
      full: { type: 'boolean', default: false },
    },
    allowPositionals: true,
  });

  const rootId = positionals[0];
  if (!rootId) {
    exitWithError('Usage: tk dep tree [--full] <id>');
  }

  const ticketMap = await buildTicketMap();
  const matchingIds = [...ticketMap.keys()].filter((id) => id.includes(rootId));

  if (matchingIds.length === 0) {
    exitWithError(`Error: ticket ${rootId} not found`);
  }
  if (matchingIds.length > 1) {
    exitWithError(`Error: ambiguous ID ${rootId}`);
  }

  const root = matchingIds[0];
  const fullMode = values.full;

  // Calculate the maximum depth at which each node appears
  const maxDepth = new Map<string, number>();
  const subtreeDepth = new Map<string, number>();
  const visited = new Set<string>();

  function findMaxDepth(id: string, currentDepth = 0, path = new Set<string>()): void {
    if (path.has(id)) return;

    maxDepth.set(id, Math.max(maxDepth.get(id) ?? 0, currentDepth));

    for (const dep of ticketMap.get(id)?.deps ?? []) {
      if (dep && !path.has(dep)) {
        findMaxDepth(dep, currentDepth + 1, new Set([...path, id]));
      }
    }
  }

  function computeSubtreeDepth(id: string): number {
    let max = maxDepth.get(id) ?? 0;
    for (const dep of ticketMap.get(id)?.deps ?? []) {
      if (dep) {
        max = Math.max(max, computeSubtreeDepth(dep));
      }
    }
    subtreeDepth.set(id, max);
    return max;
  }

  findMaxDepth(root);
  computeSubtreeDepth(root);

  function getPrintableChildren(id: string, depth: number): string[] {
    const deps = ticketMap.get(id)?.deps ?? [];
    const children = deps.filter((dep) => {
      if (!dep || !maxDepth.has(dep)) return false;
      return fullMode || depth + 1 === maxDepth.get(dep);
    });

    return children.sort((a, b) => {
      const depthDiff = (subtreeDepth.get(b) ?? 0) - (subtreeDepth.get(a) ?? 0);
      return depthDiff !== 0 ? depthDiff : a.localeCompare(b);
    });
  }

  function printTree(id: string, depth = 0, prefix = ''): void {
    const children = getPrintableChildren(id, depth);

    for (let i = 0; i < children.length; i++) {
      const child = children[i];
      const isLast = i === children.length - 1;
      const connector = isLast ? '└── ' : '├── ';
      const childPrefix = isLast ? '    ' : '│   ';
      const ticket = ticketMap.get(child);

      console.log(`${prefix}${connector}${child} [${ticket?.status ?? '?'}] ${ticket?.title ?? ''}`);

      if (!fullMode) visited.add(child);
      printTree(child, depth + 1, prefix + childPrefix);
    }
  }

  const rootTicket = ticketMap.get(root);
  console.log(`${root} [${rootTicket?.status ?? '?'}] ${rootTicket?.title ?? ''}`);
  visited.add(root);
  printTree(root);
}

export async function cmdDep(args: string[]): Promise<void> {
  if (args[0] === 'tree') {
    return cmdDepTree(args.slice(1));
  }

  const { positionals } = parseArgs({ args, allowPositionals: true });

  if (positionals.length < 2) {
    exitWithError('Usage: tk dep <id> <dep-id>\n       tk dep tree <id>  - show dependency tree');
  }

  const [id, depId] = positionals;
  const ticket = await Ticket.find(id);
  await Ticket.find(depId); // Validate dep exists

  const added = await ticket.addToArrayField('deps', depId);
  console.log(added ? `Added dependency: ${ticket.id} -> ${depId}` : 'Dependency already exists');
}

export async function cmdUndep(args: string[]): Promise<void> {
  const { positionals } = parseArgs({ args, allowPositionals: true });

  if (positionals.length < 2) {
    exitWithError('Usage: tk undep <id> <dependency-id>');
  }

  const [id, depId] = positionals;
  const ticket = await Ticket.find(id);
  const removed = await ticket.removeFromArrayField('deps', depId);

  if (!removed) {
    exitWithError('Dependency not found');
  }

  console.log(`Removed dependency: ${ticket.id} -/-> ${depId}`);
}

File: src/commands/dep.rs
use std::collections::HashSet;

use serde_json::json;

use super::CommandOutput;
use crate::commands::dep_tree::{DepthCalculator, TreeBuilder, TreeFormatter};
use crate::error::{JanusError, Result};
use crate::events::{log_dependency_added, log_dependency_removed};
use crate::ticket::{Ticket, build_ticket_map, check_circular_dependency, resolve_id_from_map};

/// Add a dependency to a ticket
pub async fn cmd_dep_add(id: &str, dep_id: &str, output_json: bool) -> Result<()> {
    let ticket = Ticket::find(id).await?;

    // Validate that the dependency exists
    let dep_ticket = Ticket::find(dep_id).await?;

    // Check for self-dependency
    if ticket.id == dep_ticket.id {
        return Err(JanusError::SelfDependency);
    }

    // Check for circular dependencies before adding
    let ticket_map = build_ticket_map().await?;
    check_circular_dependency(&ticket.id, &dep_ticket.id, &ticket_map)?;

    let added = ticket.add_to_array_field("deps", &dep_ticket.id)?;
    let metadata = ticket.read()?;

    // Log the event if dependency was actually added
    if added {
        log_dependency_added(&ticket.id, &dep_ticket.id);
    }

    let text = if added {
        format!("Added dependency: {} -> {}", ticket.id, dep_ticket.id)
    } else {
        "Dependency already exists".to_string()
    };

    CommandOutput::new(json!({
        "id": ticket.id,
        "action": if added { "dep_added" } else { "dep_already_exists" },
        "dep_id": dep_ticket.id,
        "current_deps": metadata.deps,
    }))
    .with_text(text)
    .print(output_json)
}

/// Remove a dependency from a ticket
pub async fn cmd_dep_remove(id: &str, dep_id: &str, output_json: bool) -> Result<()> {
    let ticket = Ticket::find(id).await?;

    // Resolve the dependency ID to get the full ID
    let dep_ticket = Ticket::find(dep_id).await?;

    let removed = ticket.remove_from_array_field("deps", &dep_ticket.id)?;
    if !removed {
        return Err(JanusError::DependencyNotFound(dep_ticket.id.clone()));
    }

    // Log the event
    log_dependency_removed(&ticket.id, &dep_ticket.id);

    let metadata = ticket.read()?;
    CommandOutput::new(json!({
        "id": ticket.id,
        "action": "dep_removed",
        "dep_id": dep_ticket.id,
        "current_deps": metadata.deps,
    }))
    .with_text(format!(
        "Removed dependency: {} -/-> {}",
        ticket.id, dep_ticket.id
    ))
    .print(output_json)
}

/// Display the dependency tree for a ticket
pub async fn cmd_dep_tree(id: &str, full_mode: bool, output_json: bool) -> Result<()> {
    let ticket_map = build_ticket_map().await?;

    let root = resolve_id_from_map(id, &ticket_map)?;

    let mut json_path = HashSet::new();
    let tree = TreeBuilder::build_json_tree(
        &root,
        &mut json_path,
        &ticket_map,
        &super::ticket_minimal_json_with_exists,
    );
    let json_output = json!({ "root": tree });

    if output_json {
        return CommandOutput::new(json_output).print(output_json);
    }

    let (max_depth, subtree_depth) = DepthCalculator::calculate_depths(&root, &ticket_map);

    let formatter = TreeFormatter::new(&ticket_map, &max_depth, &subtree_depth);
    formatter.print_root(&root);
    formatter.print_tree(&root, 0, "", full_mode);

    Ok(())
}
